### ISSUE TITLE
FR: In-place sanitization for nodes

### ISSUE THREAD

**Issue by choumx** [cure53/DOMPurify#288]
This issue proposes a feature which sanitizes nodes in-place instead of importing it into a new Document and returning a sanitized copy.

Sample PR: #289

### Background & Context

`DOMPurify.sanitize(node, {RETURN_DOM: true})` returns a sanitized copy of `node`. For some use cases, it would be useful if DOMPurify sanitized `node` itself instead. 

### Feature

One idea is to add a new configuration option called `IN_PLACE` that defaults to `false`. I.e.

`DOMPurify.sanitize(node, {IN_PLACE: true})`

This would be useful for incremental sanitization use cases (e.g. DOM streaming/mutation APIs) by avoiding the performance overhead of (1) deep importing/copying `node` and (2) DOM mutation caused by replacing `node` with its sanitized copy.

The specific use case we're interested in is https://github.com/ampproject/amphtml/issues/13471. We're experimenting with running untrusted JS (e.g. a React app) in a web worker containing a virtual DOM. The virtual DOM would listen for mutations and pipe changes back to the main page. We need to sanitize those mutations to prevent DOM XSS and are interested in trying out DOMPurify to do this. We recently started using DOMPurify for runtime template sanitization -- thank you!

Any thoughts or feedback would be appreciated. Also curious whether there are security implications here.

**Comment by choumx** [comment]
FYI with a quick a dirty test on TodoMVC, it looks like doing sanitization in-place offers substantial performance wins. For sanitizing the following DOM snippet (a "todo" list item):

```html
<li class="">
  <div class="view">
    <input type="checkbox" class="toggle"><label>a</label>
    <button class="destroy"></button>
  </div>
</li>
```

Out-of-place: 5-6ms
In-place: 1-2ms

**Comment by cure53** [comment]
Hey there, this sounds very interesting! We would be up for it, the performance benefits sound delicious.

However, I am interested in how you got the test data on TodoMVC. Is there a PR already we could have a look at? How did you perform the mentioned benchmark?

One of the core principles of DOMPurify is not to trust the surrounding page too much and create clean copies of important properties and methods before we even touch the user controlled markup. Is this still, something we could be able to follow with this approach?

**Comment by choumx** [comment]
Sorry for the delay. I was testing our prototype of the referenced feature on my workstation with 6x CPU throttling on Chrome. Here's an isolated lab test I quickly whipped up:

```html
<!doctype html>
<html>
  <head>
    <script type="text/javascript" src="dist/purify.min.js"></script>
    <script>
      document.addEventListener('DOMContentLoaded', function() {
        const measure = [];

        for (let i = 0; i < 1000; i++) {
          const node = document.querySelector('li');
          const start = performance.now();
          if (true) {
            DOMPurify.sanitize(node, {IN_PLACE: true});
          } else {
            const clean = DOMPurify.sanitize(node, {RETURN_DOM: true});
            node.parentElement.replaceChild(clean.firstChild, node);
          }
          measure.push(performance.now() - start);
        }

        const mean = measure.reduce((sum, val) => sum + val, 0) / measure.length;
        console.log(mean);
      }, false);
    </script>
  </head>
  <body>
    <!-- A mock todo list item. -->
    <li class="">
      <div class="view">
        <input type="checkbox" class="toggle"><label>a</label>
        <button class="destroy"></button>
      </div>
    </li>
  </body>
</html>
```

On my machine with 6x throttling, IN_PLACE takes ~0.88ms and out-of-place takes ~1.92ms per sanitization. At a glance, much of the delta appears to caused by `replaceChild` and `createHTMLDocument`.

**Comment by choumx** [comment]
Aside: In practice, we'd sanitize new nodes before they're connected to the DOM. I.e.

1. Receive struct `s` representing a new element from the web worker
2. Create Element `el` from the struct `s`
3. Sanitize `clean = sanitize(el)`
4. Append `clean` to the DOM

I'm also thinking about how to handle attribute and property changes. 

For attributes, I can do something like:

```js
const {attr, value} = attributeChangesForElement[el];
el.setAttribute(attr, value);
DOMPurify.sanitize(el, {IN_PLACE: true}); // Scary if this fails.
```

An API for validating an attribute value would help here. 😄 E.g. 

```js
const {attr, value} = attributeChangesForElement[el];
if (DOMPurify.isValidAttribute(el, attr, value)) { // More resilient.
  el.setAttribute(attr, value);
}
```

Properties are trickier. Perhaps a lookup table of property -> corresponding attribute, and only allowing the property change if the attribute change is allowed. Still working through this.

Sorry for the long digression. Thanks for considering this!

**Comment by choumx** [comment]
No worries! I'm also happy to contribute that FR if you find it acceptable. Thanks again for being so responsive and amenable.

**Comment by choumx** [comment]
#292 is a sample implementation for https://github.com/cure53/DOMPurify/issues/288#issuecomment-402299014. Would appreciate any thoughts or feedback!

**Comment by cure53** [comment]
Fantastic stuff, thanks. I just merged both PRs. When would be the right time for me to initiate a new release and edit the README?

**Comment by choumx** [comment]
I think we're in good shape now with these two features merged. I'll follow up with a unit test for `DOMPurify.isValidAttribute` soon. Thank you so much!

**Comment by cure53** [comment]
Sweet, thanks! I btw fixed one of the tests because it caused IE/Edge to throw an exception:

**Comment by choumx** [comment]
Fixed by #289 and #292.
